rmw_cyclonedds_cpp/src/serdata.cpp

```diff
@@ -685,11 +749,351 @@ static std::string get_type_name(const char * type_support_identifier, void * ty
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT16:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_UINT16);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT32:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_INT32);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT32:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_UINT32);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT64:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_INT64);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT64:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_UINT64);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING:
+        {
+          dds_dynamic_type_t ddt;
+          if (member->is_upper_bound_)
+          {
+            ddt = dds_dynamic_type_create(dds_ppant,
+              get_dynamic_type_descriptor(DDS_DYNAMIC_STRING8, nullptr, 1, (uint32_t*)(&member->array_size_), {})); 
+          } else {
+            ddt = dds_dynamic_type_create(dds_ppant, 
+              get_dynamic_type_descriptor(DDS_DYNAMIC_STRING8, nullptr, 0, nullptr, {})); 
+          }
+          if (!member->is_array_)
+          {
+            ret = dds_dynamic_type_add_member(dstruct, get_dynamic_member_descriptor(ddt, member->name_));
+            if (ret != DDS_RETCODE_OK)
+              RCUTILS_LOG_ERROR("failed to add type member: %d", DDS_DYNAMIC_STRING8);
+          } else {
+            dynamic_type_add_array(dstruct, dds_ppant, member, ddt);
+          }
+
+          break;
+        }
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_MESSAGE:
+        {
+          if (!member->members_)
+          {
+            RMW_SET_ERROR_MSG("no members value");
+            return false;
+          }
+
+          auto sub_members = static_cast<const MembersType *>(member->members_->data);
+          dds_dynamic_type_t ddt;
+          ddt = dds_dynamic_type_create(dds_ppant, 
+            get_dynamic_type_descriptor(DDS_DYNAMIC_STRUCTURE, create_type_name<MembersType>(member->members_->data).c_str(), 0, nullptr, {}));
+
+          dds_dynamic_type_set_extensibility(&ddt, DDS_DYNAMIC_TYPE_EXT_FINAL);
+
+          if (!construct_dds_dynamic_type(&ddt, dds_ppant, sub_members))
+          {
+            RMW_SET_ERROR_MSG("construct_dds_dynamic_type error");
+            return false;
+          }
+
+          if (!member->is_array_)
+          {
+            ret = dds_dynamic_type_add_member(dstruct, get_dynamic_member_descriptor(ddt, member->name_));
+            if (ret != DDS_RETCODE_OK)
+              RCUTILS_LOG_ERROR("failed to add dynamic type member");
+          } else {
+            dynamic_type_add_array(dstruct, dds_ppant, member, ddt);
```

Comment:
I'm surprised it checks for `is_array_` but apparently doesn't need to deal with `is_upper_bound_` or `array_size_`. Is that really true?

The `ROS_TYPE_STRING` case is again subtly different in this regard ...

I should check the `rosidl` definitions, I know, but for now, perhaps you can just double-check? ☺️

Reply:
This is where I made a big mistake. I didn't check `string_upper_bound`.